File: packages/shared/src/schemas/consensus-decision.ts
/**
 * ConsensusDecision Schema
 * Represents the aggregated decision from multiple agent opinions
 * 
 * @version 1.0.0
 * @backward-compatibility
 * - v1.0.0: Initial schema
 */

import { z } from "zod";
import {
  createVersionedSchema,
  sentimentSchema,
  severitySchema,
  addressSchema,
  weiAmountSchema,
  uuidSchema,
  CURRENT_SCHEMA_VERSION,
} from "./common.js";
import { recommendedActionSchema } from "./agent-opinion.js";

// ============================================
// SCHEMA DEFINITION
// ============================================

export const consensusMethodSchema = z.enum([
  "majority_vote",
  "weighted_average",
  "unanimous",
  "confidence_weighted",
  "hierarchical",
]);

export type ConsensusMethod = z.infer<typeof consensusMethodSchema>;

export const consensusApprovalStatusSchema = z.enum([
  "pending",
  "approved",
  "rejected",
  "expired",
  "auto_approved",
  "auto_rejected",
]);

export type ConsensusApprovalStatus = z.infer<typeof consensusApprovalStatusSchema>;

export const consensusDecisionSchema = createVersionedSchema({
  // Context
  contextDescription: z.string(),
  tokenAddress: addressSchema.optional(),
  tokenSymbol: z.string().optional(),
  
  // Contributing opinions
  opinionIds: z.array(uuidSchema).min(1),
  opinionCount: z.number().int().min(1),
  
  // Consensus method
  consensusMethod: consensusMethodSchema,
  consensusThreshold: z.number().min(0).max(1).default(0.6),
  consensusReached: z.boolean(),
  
  // Final decision
  finalRecommendation: recommendedActionSchema,
  finalSentiment: sentimentSchema,
  
  // Aggregated scores
  aggregatedConfidence: z.number().min(0).max(1),
  aggregatedRiskScore: z.number().min(0).max(1),
  agreementScore: z.number().min(0).max(1), // How much agents agreed
  
  // Risk assessment
  riskLevel: severitySchema,
  riskSummary: z.string(),
  
  // Reasoning
  consolidatedReasoning: z.string(),
  keyFactors: z.array(z.string()).default([]),
  disssentingViews: z.array(z.object({
    agentId: z.string(),
    view: z.string(),
    confidence: z.number().min(0).max(1),
  })).default([]),
  
  // Recommended execution parameters
  recommendedAmount: weiAmountSchema.optional(),
  recommendedAmountUsd: z.number().min(0).optional(),
  recommendedSlippage: z.number().min(0).max(100).optional(),
  
  // Approval workflow
  requiresManualApproval: z.boolean(),
  approvalStatus: consensusApprovalStatusSchema.default("pending"),
  approvedBy: z.string().optional(),
  approvedAt: z.string().datetime().optional(),
  rejectionReason: z.string().optional(),
  
  // Execution link
  executionPlanId: uuidSchema.optional(),
  
  // Timing
  decisionMadeAt: z.string().datetime(),
  expiresAt: z.string().datetime(),
});

export type ConsensusDecision = z.infer<typeof consensusDecisionSchema>;

// ============================================
// FACTORY FUNCTIONS
// ============================================

export function createConsensusDecision(
  data: Omit<ConsensusDecision, "id" | "schemaVersion" | "createdAt">
): ConsensusDecision {
  return consensusDecisionSchema.parse({
    ...data,
    id: crypto.randomUUID(),
    schemaVersion: CURRENT_SCHEMA_VERSION,
    createdAt: new Date().toISOString(),
  });
}

// ============================================
// EXAMPLES & FIXTURES
// ============================================

export const consensusDecisionExamples: ConsensusDecision[] = [
  {
    schemaVersion: "1.0.0",
    id: "550e8400-e29b-41d4-a716-446655440050",
    createdAt: "2024-01-15T14:05:00Z",
    contextDescription: "Buy decision for PEPE token based on social momentum",
    tokenAddress: "0x1234567890123456789012345678901234567890",
    tokenSymbol: "PEPE",
    opinionIds: [
      "550e8400-e29b-41d4-a716-446655440040",
      "550e8400-e29b-41d4-a716-446655440041",
      "550e8400-e29b-41d4-a716-446655440042",
    ],
    opinionCount: 3,
    consensusMethod: "confidence_weighted",
    consensusThreshold: 0.6,
    consensusReached: true,
    finalRecommendation: "buy",
    finalSentiment: "bullish",
    aggregatedConfidence: 0.78,
    aggregatedRiskScore: 0.38,
    agreementScore: 0.85,
    riskLevel: "medium",
    riskSummary: "Moderate risk due to volatility, but strong social signals support position",
    consolidatedReasoning: "Three agents agree on bullish outlook. Market analyzer sees strong momentum, sentiment analyzer confirms positive social signals, risk assessor notes acceptable risk levels.",
    keyFactors: [
      "Strong influencer activity",
      "Volume surge",
      "Positive sentiment trend",
      "Acceptable liquidity",
    ],
    disssentingViews: [],
    recommendedAmount: "100000000000000000", // 0.1 MON
    recommendedAmountUsd: 50,
    recommendedSlippage: 2.5,
    requiresManualApproval: true,
    approvalStatus: "pending",
    decisionMadeAt: "2024-01-15T14:05:00Z",
    expiresAt: "2024-01-15T14:35:00Z",
  },
];

File: services/ingestion/src/schemas/consensus_decision.rs
//! ConsensusDecision Schema
//!
//! Represents the aggregated decision from multiple agent opinions
//! Compatible with TypeScript ConsensusDecision schema

use serde::{Deserialize, Serialize};
use super::common::{Sentiment, Severity, Address, WeiAmount, Uuid, Timestamp, SchemaVersion};
use super::agent_opinion::RecommendedAction;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConsensusMethod {
    MajorityVote,
    WeightedAverage,
    Unanimous,
    ConfidenceWeighted,
    Hierarchical,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalStatus {
    Pending,
    Approved,
    Rejected,
    Expired,
    AutoApproved,
    AutoRejected,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DissentingView {
    pub agent_id: String,
    pub view: String,
    pub confidence: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConsensusDecision {
    // Base fields
    pub schema_version: SchemaVersion,
    pub id: Uuid,
    pub created_at: Timestamp,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<Timestamp>,
    
    // Context
    pub context_description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_address: Option<Address>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_symbol: Option<String>,
    
    // Contributing opinions
    pub opinion_ids: Vec<Uuid>,
    pub opinion_count: u32,
    
    // Consensus method
    pub consensus_method: ConsensusMethod,
    #[serde(default = "default_threshold")]
    pub consensus_threshold: f64,
    pub consensus_reached: bool,
    
    // Final decision
    pub final_recommendation: RecommendedAction,
    pub final_sentiment: Sentiment,
    
    // Aggregated scores
    pub aggregated_confidence: f64,
    pub aggregated_risk_score: f64,
    pub agreement_score: f64,
    
    // Risk assessment
    pub risk_level: Severity,
    pub risk_summary: String,
    
    // Reasoning
    pub consolidated_reasoning: String,
    #[serde(default)]
    pub key_factors: Vec<String>,
    #[serde(default)]
    pub disssenting_views: Vec<DissentingView>,
    
    // Recommended execution parameters
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recommended_amount: Option<WeiAmount>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recommended_amount_usd: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recommended_slippage: Option<f64>,
    
    // Approval workflow
    pub requires_manual_approval: bool,
    #[serde(default = "default_approval_status")]
    pub approval_status: ApprovalStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approved_by: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approved_at: Option<Timestamp>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rejection_reason: Option<String>,
    
    // Execution link
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_plan_id: Option<Uuid>,
    
    // Timing
    pub decision_made_at: Timestamp,
    pub expires_at: Timestamp,
}

fn default_threshold() -> f64 {
    0.6
}

fn default_approval_status() -> ApprovalStatus {
    ApprovalStatus::Pending
}

impl ConsensusDecision {
    pub fn is_actionable(&self) -> bool {
        self.consensus_reached 
            && matches!(self.approval_status, ApprovalStatus::Approved | ApprovalStatus::AutoApproved)
            && !matches!(self.final_recommendation, RecommendedAction::Hold | RecommendedAction::Avoid)
    }
    
    pub fn is_expired(&self) -> bool {
        let now = chrono::Utc::now();
        if let Ok(expires) = chrono::DateTime::parse_from_rfc3339(&self.expires_at) {
            now > expires
        } else {
            false
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_consensus_decision_typescript_compatibility() {
        let ts_json = r#"{
            "schemaVersion": "1.0.0",
            "id": "550e8400-e29b-41d4-a716-446655440050",
            "createdAt": "2024-01-15T14:05:00Z",
            "contextDescription": "Buy decision for PEPE token",
            "tokenAddress": "0x1234567890123456789012345678901234567890",
            "tokenSymbol": "PEPE",
            "opinionIds": ["550e8400-e29b-41d4-a716-446655440040"],
            "opinionCount": 3,
            "consensusMethod": "confidence_weighted",
            "consensusThreshold": 0.6,
            "consensusReached": true,
            "finalRecommendation": "buy",
            "finalSentiment": "bullish",
            "aggregatedConfidence": 0.78,
            "aggregatedRiskScore": 0.38,
            "agreementScore": 0.85,
            "riskLevel": "medium",
            "riskSummary": "Moderate risk due to volatility",
            "consolidatedReasoning": "Three agents agree on bullish outlook.",
            "keyFactors": ["Strong influencer activity"],
            "disssentingViews": [],
            "recommendedAmount": "100000000000000000",
            "requiresManualApproval": true,
            "approvalStatus": "pending",
            "decisionMadeAt": "2024-01-15T14:05:00Z",
            "expiresAt": "2024-01-15T14:35:00Z"
        }"#;

        let parsed: ConsensusDecision = serde_json::from_str(ts_json).unwrap();
        
        assert_eq!(parsed.consensus_method, ConsensusMethod::ConfidenceWeighted);
        assert_eq!(parsed.final_recommendation, RecommendedAction::Buy);
        assert!(parsed.consensus_reached);
        assert!(!parsed.is_actionable()); // pending approval
    }
}
